File: daily_problems/2025/02/0224/personal_submission/cf222c_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divide, I conquer!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int M = 1e7;
    vector<int> prime_factor(M + 1);
    iota(prime_factor.begin(), prime_factor.end(), 0);

    for (int i = 2; i <= M; i ++) {
        if (prime_factor[i] == i) {
            for (int j = i; j <= M; j += i) {
                prime_factor[j] = i;
            }
        }
    }

    int n, m;
    cin >> n >> m;

    vector<int> nums1(n), nums2(m);
    for (auto &x: nums1) cin >> x;
    for (auto &x: nums2) cin >> x;

    vector<int> cnt1(M + 1, 0), cnt2(M + 1, 0);

    for (int i = 0; i < n; i ++) {
        int x = nums1[i];
        while (x > 1) {
            cnt1[prime_factor[x]] ++;
            x /= prime_factor[x];
        }
    }

    for (int i = 0; i < m; i ++) {
        int x = nums2[i];
        while (x > 1) {
            cnt2[prime_factor[x]] ++;
            x /= prime_factor[x];
        }
    }

    for (int i = 0; i <= M; i ++) {
        cnt1[i] = min(cnt1[i], cnt2[i]);
        cnt2[i] = cnt1[i];
    }

    for (int i = 0; i < n; i ++) {
        int x = nums1[i];
        while (x > 1) {
            if (cnt1[prime_factor[x]]) {
                cnt1[prime_factor[x]] --;
                nums1[i] /= prime_factor[x];
            }
            x /= prime_factor[x];
        }
    }

    for (int i = 0; i < m; i ++) {
        int x = nums2[i];
        while (x > 1) {
            if (cnt2[prime_factor[x]]) {
                cnt2[prime_factor[x]] --;
                nums2[i] /= prime_factor[x];
            }
            x /= prime_factor[x];
        }
    }

    cout << n << ' ' << m << '\n';
    for (auto &v: nums1) cout << v << ' ';
    cout << '\n';
    for (auto &v: nums2) cout << v << ' ';
    cout << '\n';

    return 0;
}

File: daily_problems/2025/02/0224/personal_submission/cf222c_yawn_sean.py
# Submission link: https://codeforces.com/contest/222/submission/307471915
def main():
    prime_factor = list(range(10 ** 7 + 1))
    for i in range(2, 10 ** 7 + 1):
        if prime_factor[i] == i:
            for j in range(i, 10 ** 7 + 1, i):
                prime_factor[j] = i

    n, m = MII()
    nums1 = LII()
    nums2 = LII()

    cnt1 = [0] * (10 ** 7 + 1)
    cnt2 = [0] * (10 ** 7 + 1)

    for x in nums1:
        while x > 1:
            cnt1[prime_factor[x]] += 1
            x //= prime_factor[x]

    for x in nums2:
        while x > 1:
            cnt2[prime_factor[x]] += 1
            x //= prime_factor[x]

    for i in range(10 ** 7 + 1):
        cnt1[i] = fmin(cnt1[i], cnt2[i])
        cnt2[i] = cnt1[i]

    for i in range(n):
        x = nums1[i]
        while x > 1:
            if cnt1[prime_factor[x]]:
                cnt1[prime_factor[x]] -= 1
                nums1[i] //= prime_factor[x]
            x //= prime_factor[x]

    for i in range(m):
        x = nums2[i]
        while x > 1:
            if cnt2[prime_factor[x]]:
                cnt2[prime_factor[x]] -= 1
                nums2[i] //= prime_factor[x]
            x //= prime_factor[x]

    print(n, m)
    print(*nums1)
    print(*nums2)